deps/0011-backend-tokenization-fallback.md

```diff
@@ -0,0 +1,222 @@
+        -> anyhow::Result<Vec<u32>>
+    {
+        Python::with_gil(|py| {
+            let result = self.py_tokenizer.call_method1(
+                py, "tokenize",
+                (messages_to_pylist(py, messages)?, model, tools.map(|t| tools_to_pylist(py, t)).transpose()?)
+            )?;
+            Ok(result.extract(py)?)
+        })
+    }
+}
+
+// Helper functions convert Rust structs to Python dicts/lists
+```
+
+#### Integration
+
+```rust
+impl OpenAIPreprocessor {
+    pub async fn preprocess_with_adapter(
+        &self,
+        request: &NvCreateChatCompletionRequest,
+        tokenizer: &dyn TokenizerAdapter,
+    ) -> Result<PreprocessedRequest> {
+        let mut builder = self.builder(request)?;  // Rust: extract params
+        let token_ids = tokenizer.tokenize(...)?;   // Python: tokenize
+        builder.token_ids(token_ids);
+        self.gather_multi_modal_data(request, &mut builder).await?;
+        builder.build()
+    }
+}
+```
+
+#### Initialization
+
+```rust
+// Rust passes tokenizer path to Python factory
+let tokenizer_path = card.get_tokenizer_path()?;
+let adapter = Python::with_gil(|py| {
+    let py_tokenizer = factory.call1(py, (tokenizer_path,))?;
+    Ok(Arc::new(PythonTokenizerAdapter::new(py_tokenizer)))
+})?;
+```
+
+```python
+# Python factory creates tokenizer from path
+def create_vllm_tokenizer(tokenizer_path: str) -> TokenizerProtocol:
+    tokenizer = AutoTokenizer.from_pretrained(os.path.dirname(tokenizer_path))
+    return VllmTokenizer(tokenizer)
+
+# Main
+tokenizer_factory = create_vllm_tokenizer if flags.vllm_processing else None
+await run_input(runtime, mode, engine, tokenizer_factory)
+```
+
+## Post-Processing
+
+### Current Architecture
+
+**Frontend**: Post-processing (detokenization, tool calling) happens in Rust via `backend.backward_edge()` and `preprocessor_op.backward_edge()`
+
+**Backend** (`serve_endpoint`): Minimal - just wraps `PythonAsyncEngine` in ingress handler
+
+```rust
+// Frontend pipeline includes Rust post-processing
+.link(service_backend)?
+.link(backend.backward_edge())?          // Detokenization in Rust
+.link(preprocessor_op.backward_edge())?  // Tool calling in Rust
+```
+
+### Two Processing Modes
+
+#### Mode 1: Rust Post-Processing on Backend (Proposed)
+
+Move Rust detokenization/tool calling to backend servers (better CPU utilization):
+
+```rust
+// Backend serve_endpoint builds larger pipeline
+let pipeline = frontend
+    .link(tool_calling.forward_edge())?  // passthrough
+    .link(backend.forward_edge())?       // passthrough
+    .link(service_backend)?              // PythonAsyncEngine
+    .link(backend.backward_edge())?      // Detokenization in Rust (on backend)
+    .link(tool_calling.backward_edge())? // Tool calling in Rust (on backend)
+    .link(frontend)?
+```
+
+**Benefits**: Offloads CPU work to backend servers, scales with engine instances
+
+#### Mode 2: vLLM Native Processing (`--vllm-processing`)
```

Comment:
Can we think of this in terms of a "hello world backend" example, CLI args, and all - and see if it generalizes to vllm, sglang, trtllm, etc. that each implement the "interface" we defined out of the box for users of our officially supported backends?

For example, maybe any backend would accept args like `--dyn-backend-preprocessor` and `--dyn-backend-postprocessor` (subject to change, just following `--dyn-` prefix convention for dynamo specific args rather than framework args we passthrough), etc. and for vllm backend that happens to map to using vllm's tokenizer?